core/arch/arm/kernel/user_ta.c

```diff
@@ -161,17 +179,64 @@ static TEE_Result load_elf_segments(struct user_ta_ctx *utc,
 			return res;
 
 		if (type == PT_LOAD) {
-			mattr = elf_flags_to_mattr(flags, init_attrs);
-			res = tee_mmu_map_add_segment(utc, utc->mobj_code,
-						      offs, size, mattr);
-			if (res != TEE_SUCCESS)
-				return res;
+			void *p = realloc(segs, (num_segs + 1) * sizeof(*segs));
+
+			if (!p) {
+				free(segs);
+				return TEE_ERROR_OUT_OF_MEMORY;
+			}
+			segs = p;
+			segs[num_segs].offs = ROUNDDOWN(offs, SMALL_PAGE_SIZE);
+			segs[num_segs].oend = ROUNDUP(offs + size,
+						      SMALL_PAGE_SIZE);
+			segs[num_segs].prot = elf_flags_to_mattr(flags,
+								 init_attrs);
+			num_segs++;
 		} else if (type == PT_ARM_EXIDX) {
 			utc->exidx_start = offs;
 			utc->exidx_size = size;
 		}
 	}
 
+	/* Sort in case load sections aren't in order */
+	qsort(segs, num_segs, sizeof(*segs), cmp_load_seg);
+
+	idx = 1;
+	while (idx < num_segs) {
+		size_t this_size = segs[idx].oend - segs[idx].offs;
+		size_t prev_size = segs[idx - 1].oend - segs[idx - 1].offs;
+
+		if (core_is_buffer_intersect(segs[idx].offs, this_size,
```

Comment:
check attributes are compatible? expected, by construction (only `assert()`) or panic if not?

Reply:
This is unchanged behavior.

The check doesn't belong here, but if needed we could add something in `vm_map()`.

However, I'm not sure we gain much more than an explicit error message instead of an unhandled prefetch abort in user mode. What's compatible settings depends on the value of `CFG_HWSUPP_MEM_PERM_WXN` and `CFG_CORE_RWDATA_NOEXEC`.

Reply:
ok. thanks for the details.